drivers/stepper/stepper_motion_controllers/zephyr_stepper_motion_controller.c

```diff
@@ -0,0 +1,405 @@
+{
+	const struct stepper_motion_control_config *config = dev->config;
+	struct stepper_motion_control_data *data = dev->data;
+
+	update_remaining_steps(dev);
+	stepper_step(config->stepper);
+	update_actual_position(dev);
+
+	if (config->timing_source->needs_reschedule(dev) && data->step_count != 0) {
+		config->timing_source->start(&config->timing_config, &data->timing_data);
+	} else if (data->step_count == 0) {
+		stepper_trigger_callback(data->dev, STEPPER_EVENT_STEPS_COMPLETED);
+		config->timing_source->stop(&config->timing_config, &data->timing_data);
+	}
+}
+
+static void velocity_mode_task(const struct device *dev)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+	struct stepper_motion_control_data *data = dev->data;
+
+	stepper_step(config->stepper);
+	update_actual_position(dev);
+
+	if (config->timing_source->needs_reschedule(dev)) {
+		(void)config->timing_source->start(&config->timing_config, &data->timing_data);
+	}
+}
+
+static int z_stepper_motion_control_set_event_callback(const struct device *dev,
+						       stepper_event_callback_t cb, void *user_data)
+{
+	struct stepper_motion_control_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		data->callback = cb;
+	}
+	data->event_cb_user_data = user_data;
+	return 0;
+}
+
+void stepper_handle_timing_signal(const struct device *dev)
+{
+	struct stepper_motion_control_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		switch (data->run_mode) {
+		case STEPPER_RUN_MODE_POSITION:
+			position_mode_task(dev);
+			break;
+		case STEPPER_RUN_MODE_VELOCITY:
+			velocity_mode_task(dev);
+			break;
+		default:
+			LOG_WRN("Unsupported run mode: %d", data->run_mode);
+			break;
+		}
+	}
+}
+
+static int z_stepper_motion_control_enable(const struct device *dev)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+
+	return stepper_enable(config->stepper);
+
+	return 0;
+}
+
+static int z_stepper_motion_control_disable(const struct device *dev)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+
+	return stepper_disable(config->stepper);
+}
+
+static int z_stepper_motion_control_set_micro_step_res(const struct device *dev,
+						       const enum stepper_micro_step_resolution res)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+
+	return stepper_set_micro_step_res(config->stepper, res);
+}
+
+static int z_stepper_motion_control_get_micro_step_res(const struct device *dev,
+						       enum stepper_micro_step_resolution *res)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+
+	int ret = stepper_get_micro_step_res(config->stepper, res);
+
+	return ret;
+}
+
+static int stepper_motion_control_init(const struct device *dev)
+{
+	const struct stepper_motion_control_config *config = dev->config;
+	struct stepper_motion_control_data *data = dev->data;
+	int ret;
+
+	data->dev = dev;
+
+	if (config->timing_source->init) {
```

Comment:
we should return an error if `config->timing_source->init` evaluates to `false` because in this case the controller wont be usable (and we have the worker timing source as fallback for that reason).

Reply:
resolved